### PR TITLE
feat: adaptive intelligence — background review, memory, skills

### PR DESCRIPTION
**Pull request by michaelneale**
## Summary

[Hermes-agent](https://hermes-agent.nousresearch.com/) is pretty cool (hermes itself is both in some ways a nicer openclaw, but also has some RL trained models on llama, but underneath it - a pretty cool agent harness which has a done of good things in it). 

Inspired by that, this brings over some ideas which seem relevant to goose, maintaining ongoing user preferences and recent memory, and curating skills, adding and adapting skills automatically, and more efficient context use. 

This shows up as a platform extension which changes goose when it is on (only).
basically - as you use this, the more personal/smart it gets.

### Core Features
- **Background memory review** — after N user turns, autonomously reviews conversation for memory-worthy facts (user prefs, env details, corrections)
- **Background skill review** — after N tool iterations (complex work), checks if a skill should be created/updated
- **Pre-compression flush** — before context compaction, saves anything worth keeping permanently
- **Cross-session pattern detection** — queries past sessions to notice recurring themes
- **MEMORY.md + USER.md format** used to store things in easily readable way
- **create_skill + patch_skill tools** — agent can create/update its own procedural skills
- **Knowledge management guidance** in system prompt

### Context Management
- Lower compaction threshold to 50%
- Proactive tool output pruning
- a more Structured compaction template
- Head+tail truncation for large tool results (shell, ACP file reads, automation scripts)

### Tests
- 30+ adaptive_memory unit tests (add/replace/remove, budgets, security, duplicates, section detection)
- 10 knowledge_review tests (tool filtering, prompt selection, trigger thresholds)
- Skill test cleanup with RAII guards

### CODE REVIEW

File: crates/goose/src/context_mgmt/mod.rs [hunk 1]
```diff
@@ -126,8 +148,13 @@ pub async fn compact_messages(
...
     let messages_to_compact = messages.as_slice();
 
-    let (summary_message, summarization_usage) =
-        do_compact(provider, session_id, messages_to_compact).await?;
+    let (summary_message, summarization_usage) = do_compact(
+        provider,
+        session_id,
+        messages_to_compact,
+        extension_manager.is_some(),
+    )
```
[the remark this answers is not in the document]
Comment (michaelneale) [reply crates/goose/src/context_mgmt/mod.rs:167]: probably valid problem yes...

**Comment by michaelneale** [comment]
closing this - used it for a while, it can overfit things.

**Comment by codefromthecrypt** [comment]
for whoever tries to pick this up next, some feedback cc @alexhancock @DOsinga as this is ACP related.

None of what I am saying was written or edited by AI. I'm weaning off using AI in general for comments ;)

So, some weeks ago I looked deeply at this and tried it locally and found some things. I was excited about skills generation but got lost in the weeds. Some of those weeds are relevant generally some are only about skills. To be fair, I spent most time trying to get the skill generation to actually trigger, how useful the output is... that I didn't get to. It is actually a hard to manage system and most is not about the approach here, rather scoping.

1. *Doing anything async bounded to a session is problematic currently*. In goose, we currently don't have a "session local" concept, especially for work. A good example of this is the "session rename task". At last look (maybe changed since), a rename can outrun the session it is about, and fall on the floor. In the case of a rename it makes sense. Potentially valuable work like deriving skills... harder to say. We don't have a clear bubble of async work to tie it back to the owning session, nor a way to prioritize work in progress at an exit boundary.
2. *ACP session boundary is not strictly closed over either*. Relating to above, if you have a session in ACP, work derived from it is zombied for the same reason. While `session/close` is now standard practice, it isn't fully converged. Even if we bounded things, we need clients to actually call session/close and decide to block or not. ACP had a `nes/close` (yeah it isn't mario, but I though that) and now `session/close`. I raised a [PR](https://github.com/openclaw/acpx/pull/283) to address huggingface/acpx and they worked it in a better way than my first guess, so it is plugged there. TL;DR; is even if we did a session-scoped boundary for work, clients need to be backedfilled.
3. *skill writing is more like eval than anything else, so we'd need a model toggle*. As much as I hate the million model config thing, it is purely the case that a fast model is a terrible choice for skill writing. There is an irony is you need more like a LLM as a Judge model, yet there is also eval concerns of the skill you are writing. If we need a model, it is likely to need to be more like an eval model. GOOSE_SKILL_MODEL/GOOSE_SKILL_CONTEXT_LIMIT
4. *if we write skills it needs to be more like a CEL expression*. It is impossible to think that various skills will be quality with a single solution. that said, it we want a paved path. If a skiller is triggered we have a choice of potentially many config (I wrote GOOSE_SKILL_REVIEW_ITERATIONS), or an expression language. I feel like the latter is more viable for the project, even if it pushes things to the user.
5. *If we did all this, the prompt is still super sensitive, run a test!*: I found all the worst edge cases by simply using ollama qwen3.. found the race conditions, skill gap, prompt problems (like faking not actually wirting a skill). we need a super tight eval.


There are things above that are tactical and strategic. I think the tactical ones some of them apply to things outside skill writing. The strategic hint at an extension problem. The bounded context of skill writing deserves its own focus area, ideally in a way that drives requirements, not impl on goose core.

Hope this helps!
